src/vanilla.ts

```diff
@@ -43,240 +15,296 @@ type Op =
-  }
-  const found = proxyCache.get(initialObject) as T | undefined
-  if (found) {
-    return found
-  }
-  let version = globalVersion
-  const listeners = new Set<Listener>()
-  const notifyUpdate = (op: Op, nextVersion = ++globalVersion) => {
-    if (version !== nextVersion) {
-      version = nextVersion
-      listeners.forEach((listener) => listener(op, nextVersion))
+  refSet = new WeakSet(),
+
+  canProxy = (x: unknown) =>
+    isObject(x) &&
+    !refSet.has(x) &&
+    (Array.isArray(x) || !(Symbol.iterator in x)) &&
+    !(x instanceof WeakMap) &&
+    !(x instanceof WeakSet) &&
+    !(x instanceof Error) &&
+    !(x instanceof Number) &&
+    !(x instanceof Date) &&
+    !(x instanceof String) &&
+    !(x instanceof RegExp) &&
+    !(x instanceof ArrayBuffer),
+
+  ref = <T extends object>(o: T): T & AsRef => {
+    refSet.add(o)
+    return o as T & AsRef
+  },
+
+  VERSION = __DEV__ ? Symbol('VERSION') : Symbol(),
+  LISTENERS = __DEV__ ? Symbol('LISTENERS') : Symbol(),
+  SNAPSHOT = __DEV__ ? Symbol('SNAPSHOT') : Symbol(),
+  PROMISE_RESULT = __DEV__ ? Symbol('PROMISE_RESULT') : Symbol(),
+  PROMISE_ERROR = __DEV__ ? Symbol('PROMISE_ERROR') : Symbol(),
+
+  proxyCache = new WeakMap<object, ProxyObject>(),
+
+  snapshotCache = new WeakMap<object, [version: number, snapshot: unknown]>(),
+
+  createSnapshot = <T extends object>(
+    version: number,
+    target: T,
+    receiver: any
+  ): T => {
+    const cache = snapshotCache.get(receiver)
+    if (cache?.[0] === version) {
+      return cache[1] as T
     }
-  }
-  const propListeners = new Map<string | symbol, Listener>()
-  const getPropListener = (prop: string | symbol) => {
-    let propListener = propListeners.get(prop)
-    if (!propListener) {
-      propListener = (op, nextVersion) => {
-        const newOp: Op = [...op]
-        newOp[1] = [prop, ...(newOp[1] as Path)]
-        notifyUpdate(newOp, nextVersion)
+    const snapshot: any = Array.isArray(target)
+      ? []
+      : Object.create(Object.getPrototypeOf(target))
+    markToTrack(snapshot, true) // mark to track
+    snapshotCache.set(receiver, [version, snapshot])
+    Reflect.ownKeys(target).forEach((key) => {
+      const value = Reflect.get(target, key, receiver)
+      if (refSet.has(value)) {
+        markToTrack(value, false) // mark not to track
+        snapshot[key] = value
+      } else if (value instanceof Promise) {
+        if (PROMISE_RESULT in value) {
+          snapshot[key] = (value as any)[PROMISE_RESULT]
+        } else {
+          const errorOrPromise = (value as any)[PROMISE_ERROR] || value
+          Object.defineProperty(snapshot, key, {
+            get() {
+              if (PROMISE_RESULT in value) {
+                return (value as any)[PROMISE_RESULT]
+              }
+              throw errorOrPromise
+            },
+          })
+        }
+      } else if (value?.[SNAPSHOT]) {
+        snapshot[key] = value[SNAPSHOT]
+      } else {
+        snapshot[key] = value
```

Comment:
Is this change from `value?.[LISTENERS]` to `value?.[SNAPSHOT]` intentional? 

- Previously, it was checking if the value was a valtio proxy -- since valtio proxies return a `Set` for `Listeners` and if true, then call `SNAPSHOT`. Maybe there should be a way, that given a valtio proxy, you're able to tell if it's a valtio proxy from the outside i.e. `value?.[VALTIO_STATE]`.
- Not sure when `value[SNAPSHOT]` will return `undefined` or `null` in `createSnapshot`. It is possible custom `snapshot` method will intentionally return `undefined` or `null` too.
- Checking if `value?.[SNAPSHOT]` is undefined, then assigning it to `snapshot[key]` seems like redundant computation as well.

Reply:
Oh! good point. It was intentional, but wrong way in #533.
I'll revert it in this PR.
btw, snapshot will always be an object.